src/byte_record.rs

```diff
@@ -362,18 +362,53 @@ impl ByteRecord {
     /// assert_eq!(record, vec!["", "foo", "bar", "b a z"]);
     /// ```
     pub fn trim(&mut self) {
+        // early return in case of no records
         let length = self.len();
         if length == 0 {
             return;
         }
-        // TODO: We could likely do this in place, but for now, we allocate.
-        let mut trimmed =
-            ByteRecord::with_capacity(self.as_slice().len(), self.len());
-        trimmed.set_position(self.position().cloned());
-        for field in &*self {
-            trimmed.push_field(field.trim());
+
+        // early return if last record ends at 0 -> all empty
+        if self.0.bounds.ends[length - 1] == 0 {
+            return;
+        };
+
+        let mut read_pos: usize = 0;
+        let mut write_pos: usize = 0;
+        for element_end_idx in &mut self.0.bounds.ends[..length] {
+            // left trim
+            let mut found_only_whitespace = true;
+            while read_pos < *element_end_idx {
```

Comment:
I'd suggest finding the first non-whitespace position first and then using [slice.copy_within](https://doc.rust-lang.org/std/primitive.slice.html#method.copy_within) to copy the rest. That would (a) make this somewhat easier to read (e.g. obviate the need for found_only_whitespace) and (b) be potentially easier to optimize for the compiler (no need to deduce that once the branch with the copy is taken it will be taken until the loop finishes).